design/one-pager-build-with-earthly.md

```diff
@@ -0,0 +1,248 @@
+# Build with Earthly
+
+* Owner: Nic Cope (@negz)
+* Status: Proposed
+
+## Background
+
+Crossplane uses a `Makefile` with several targets, like `make build`, to
+automate tasks that developers frequently need to run when developing
+Crossplane.
+
+Crossplane also uses GitHub Actions for continous integration (CI), to validate
+pull requests. Most of Crossplane's GitHub Actions workflows run the same Make
+targets. This creates some consistency between local development and CI. For
+example `make test` should have the same result whether run locally or in CI.
+
+The `Makefile` includes a moderate library of other `Makefiles`. These are
+imported from the `build/makelib` directory. The `build` directory is a Git
+submodule. Its source is https://github.com/crossplane/build. Most maintainers
+call it "the build submodule".
+
+Crossplane uses the build submodule to:
+
+- Install pinned versions of common tools (`helm`, `kubectl`, etc)
+- Cross-compile Crossplane for several platforms
+- Produce a multi-platform OCI image for Crossplane
+- Run code generation - e.g. `go generate`
+- Validate code by running linters, unit tests, and end-to-end (E2E) tests
+- Automatically derive the semantic version of artifacts from git tags
+- Publish OCI image artifacts to OCI registries
+- Publish binary and Helm chart artifacts to AWS S3
+
+The build submodule is also used by Crossplane extensions, like Providers.
+Providers use the build submodule to do more than core Crossplane - for example
+they use it to spin up `kind` clusters and deploy Crossplane for testing.
+
+In the 5+ years I've been a Crossplane maintainer, almost every new maintainer
+(including myself) has expressed a dislike for the build submodule and a desire
+to change build tooling.
+
+I believe folks dislike the build submodule because:
+
+- Make, as a language, has a high learning curve
+- Few people have prior experience with advanced use of Make
+- Needing to update a shared git submodule slows down changes to build logic
```

Comment:
> - Needing to update a shared git submodule slows down changes to build logic

This indeed prevents some changes in build logic from unintentionally breaking builds in another repository. I feel much more comfortable making changes in the build repo since I know it won't impact anyone until they intentionally consume my change (where they can validate before merging the PR).

Reply:
That's true, but I could also see it creating issues as well.

e.g. I make a change, then weeks or months later you finally upgrade to start consuming the change (maybe because you wanted some _other_ change) and now you're having to figure out and deal with the impact of my change.

At first glance it _seems_ like Earthly is straightforward enough that I wouldn't expect to need to use its `IMPORT` functionality. I'd hope each repo could just have their own `Earthfile`, even if certain targets (e.g. `+build`) were pretty same-y. This way each repo can have as simple or as complex an `Earthfile` as they need. For example crossplane-runtime wouldn't need most of the targets this repo does.